File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/StringTupleDeserializer.java
package it.unibo.coordination.linda.presentation;

import com.fasterxml.jackson.databind.ObjectMapper;
import it.unibo.coordination.linda.text.StringTuple;

import java.util.Map;

class StringTupleDeserializer extends DynamicDeserializer<StringTuple> {

    public StringTupleDeserializer(MIMETypes mimeType, ObjectMapper mapper) {
        super(StringTuple.class, mimeType, mapper);
    }

    @Override
    public StringTuple fromDynamicObject(Object dynamicObject) {
        if (dynamicObject instanceof Map) {
            var map = (Map<String, ?>) dynamicObject;
            if (map.containsKey("tuple")) {
                var tuple = map.get("tuple");
                if (tuple instanceof String) {
                    return StringTuple.of((String) tuple);
                }
            }
        }
        throw new IllegalArgumentException("Cannot read " + getSupportedMIMEType());
    }
}

File: linda-text-presentation/src/main/kotlin/it/unibo/coordination/linda/text/StringTupleDeserializer.kt
package it.unibo.coordination.linda.text

import com.fasterxml.jackson.databind.ObjectMapper
import it.unibo.presentation.DynamicDeserializer
import it.unibo.presentation.MIMETypes

internal class StringTupleDeserializer(mimeType: MIMETypes, mapper: ObjectMapper) : DynamicDeserializer<StringTuple>(StringTuple::class.java, mimeType, mapper) {
    override fun fromDynamicObject(dynamicObject: Any): StringTuple {
        if (dynamicObject is Map<*, *>) {
            if (dynamicObject.containsKey("tuple")) {
                val tuple = dynamicObject["tuple"]
                if (tuple is String) {
                    return StringTuple.of(tuple)
                }
            }
        }
        throw IllegalArgumentException("Cannot read $supportedMIMEType")
    }
}